### PR TITLE
store applied filters in onyx

### PR DESCRIPTION
**Pull request by 289Adam289**
### Fixed Issues

$ https://github.com/Expensify/App/issues/48156
$ https://github.com/Expensify/App/issues/48407
PROPOSAL:

### Tests

1. Open the app
2. Go to Search
3. From the available options, select both a Type (Expense, Invoice, Trip, or Chat) and a Status (from the status bar).
4. Click "Filters" button
5. Verify that the available filters are relevant to the selected type
6. Apply filters as needed
7. Close the filters page
8. Click "Filters" button
9. Verify that the previous filters values are selected
10. Click on "View results"
11. Verify that the displayed results match the previously selected Type and Status
12. Click "Filters" button
13. Verify that the previously used filters are selected

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open the app
2. Go to Search
3. From the available options, select both a Type (Expense, Invoice, Trip, or Chat) and a Status (from the status bar).
4. Click "Filters" button
5. Verify that the available filters are relevant to the selected type
6. Apply filters as needed
7. Close the filters page
8. Click "Filters" button
9. Verify that the previous filters values are selected
10. Click on "View results"
11. Verify that the displayed results match the previously selected Type and Status
12. Click "Filters" button
13. Verify that the previously used filters are selected

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/3231cc3a-47c8-403a-83b4-3b3b60aa4e61

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/5e821da0-280f-4e9b-8172-02d12bd349f0

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/eb8c31e3-51d6-4665-a3b6-374bd2b5c7f3
</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/8fc20ef0-0bf2-408d-9ddc-a3407bba6dac
</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/ca63b2ce-1585-441e-a547-80b49742a248

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by Guccio163 — APPROVED** [review]
LGTM when it comes to strictly this PR, but I found bug in `Category filter` and it's probably linked to parser, maybe you can take a look at it @289Adam289 :

- If you create `Category` with `&amp;` written inside, select this category in `Filters`, save and display results, the whole page crashed:

https://github.com/user-attachments/assets/a1e7924c-eb2b-4782-b6f6-a67639b56a5a

Additionally, it doesn't happen if I create tag with `&` alone in it, which encoded version in `&amp;`

File: src/libs/SearchUtils.ts [hunk 1]
```diff
@@ -529,6 +537,41 @@ function getFilters(queryJSON: SearchQueryJSON) {
     return filters;
 }
 
+function getFiltersForm(filters: QueryFilters) {
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:540]:
```suggestion
function getFiltersFormValues(filters: QueryFilters) {
```

File: src/components/Search/SearchPageHeader.tsx [hunk 2]
```diff
@@ -314,7 +314,14 @@ function SearchPageHeader({queryJSON, hash, onSelectDeleteOption, setOfflineModa
             <Button
                 text={translate('search.filtersHeader')}
                 icon={Expensicons.Filters}
-                onPress={() => Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS)}
+                onPress={() => {
+                    const filters = SearchUtils.getFilters(queryJSON);
+                    const form = SearchUtils.getFiltersForm(filters);
+                    form.type = queryJSON.type;
+                    form.status = queryJSON.status;
```
Comment (Kicu) [inline src/components/Search/SearchPageHeader.tsx:321]: you created `. getFiltersForm()` method to use in this place specifically - so it should do all the work needed for creating form data.
Because of that it should set `form.type` and `status` inside itself, so that when you call it, all the data is prepared.

You can either pass it `type` and `status` as separate arguments, or just pass whole `queryJSON` as the 2nd param - I think both are ok.

File: src/components/Search/SearchPageHeader.tsx [hunk 3]
```diff
@@ -314,7 +314,14 @@ function SearchPageHeader({queryJSON, hash, onSelectDeleteOption, setOfflineModa
             <Button
                 text={translate('search.filtersHeader')}
                 icon={Expensicons.Filters}
-                onPress={() => Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS)}
+                onPress={() => {
+                    const filters = SearchUtils.getFilters(queryJSON);
+                    const form = SearchUtils.getFiltersForm(filters);
```
Comment (Kicu) [inline src/components/Search/SearchPageHeader.tsx:319]:
```suggestion
                    const values = SearchUtils.getFiltersForm(filters);
```

Totally not a bug, but I checked in the code and we almost always call `updateAdvancedFilters` with the argument `values` so I thought maybe we should always refer to this object as values 🤷

File: src/libs/SearchUtils.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ import * as TransactionUtils from './TransactionUtils';
 import * as UserUtils from './UserUtils';
 
 type KeysOfFilterKeysObject = keyof typeof CONST.SEARCH.SYNTAX_FILTER_KEYS;
+type KeysOfRootKeysObject = keyof typeof CONST.SEARCH.SYNTAX_ROOT_KEYS;
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:25]: Please rename both to more simple: `FilterKeys`, `RootKeys` - this "keysOfKeys...object" naming is redundant IMO

File: src/libs/SearchUtils.ts [hunk 5]
```diff
@@ -451,7 +452,14 @@ function buildQueryStringFromFilters(filterValues: Partial<SearchAdvancedFilters
         if (filterKey === FILTER_KEYS.KEYWORD && filterValue) {
             const keyInCorrectForm = (Object.keys(CONST.SEARCH.SYNTAX_FILTER_KEYS) as KeysOfFilterKeysObject[]).find((key) => CONST.SEARCH.SYNTAX_FILTER_KEYS[key] === filterKey);
             if (keyInCorrectForm) {
-                return `${CONST.SEARCH.SYNTAX_FILTER_KEYS[keyInCorrectForm]}:${filterValue as string}`;
+                return `${filterValue as string}`;
+            }
+        }
+
+        if ((filterKey === FILTER_KEYS.TYPE || filterKey === FILTER_KEYS.STATUS) && filterValue) {
+            const keyInCorrectForm = (Object.keys(CONST.SEARCH.SYNTAX_ROOT_KEYS) as KeysOfRootKeysObject[]).find((key) => CONST.SEARCH.SYNTAX_ROOT_KEYS[key] === filterKey);
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:460]: we don't need this `find`, simply do an if case for `TYPE` and for `STATUS` separately

File: src/pages/Search/SearchTypeMenuNarrow.tsx [hunk 6]
```diff
@@ -96,7 +100,14 @@ function SearchTypeMenuNarrow({typeMenuItems, activeItemIndex, title}: SearchTyp
             </PressableWithFeedback>
             <Button
                 icon={Expensicons.Filters}
-                onPress={() => Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS)}
+                onPress={() => {
+                    const filters = SearchUtils.getFilters(queryJSON);
+                    const form = SearchUtils.getFiltersForm(filters);
+                    form.type = queryJSON.type;
+                    form.status = queryJSON.status;
```
Comment (Kicu) [inline src/pages/Search/SearchTypeMenuNarrow.tsx:107]: same comment as in the other file

**Review by Kicu — CHANGES_REQUESTED** [review]
In general code looks good, I have left mostly minor comments about naming, and 1 comment to simplify code.
Good job 👍

**Comment by 289Adam289** [comment]
@luacmartins ready for review

File: src/pages/Search/AdvancedSearchFilters.tsx [hunk 7]
```diff
@@ -184,9 +186,6 @@ function getFilterDisplayTitle(filters: Partial<SearchAdvancedFiltersForm>, fiel
         return filters[fieldName];
     }
 
-    // Todo Once all Advanced filters are implemented this line can be cleaned up. See: https://github.com/Expensify/App/issues/45026
-    // @ts-expect-error this property access is temporarily an error, because not every SYNTAX_FILTER_KEYS is handled by form.
-    // When all filters are updated here: src/types/form/SearchAdvancedFiltersForm.ts this line comment + type cast can be removed.
     const filterValue = filters[fieldName] as string;
```
Comment (Kicu) [inline src/pages/Search/AdvancedSearchFilters.tsx:189]: So in the end can we remove `as string` or not? 
and if not can we do anything better than casting this value with `as`?

File: src/libs/SearchUtils.ts [hunk 8]
```diff
@@ -469,8 +471,8 @@ function buildQueryStringFromFilters(filterValues: Partial<SearchAdvancedFilters
             Array.isArray(filterValue) &&
             filterValue.length > 0
         ) {
-            const filterValueArray = filterValues[filterKey] ?? [];
-            const keyInCorrectForm = (Object.keys(CONST.SEARCH.SYNTAX_FILTER_KEYS) as KeysOfFilterKeysObject[]).find((key) => CONST.SEARCH.SYNTAX_FILTER_KEYS[key] === filterKey);
+            const filterValueArray = Array.from(new Set<string>(filterValues[filterKey] ?? []));
```
Comment (Kicu) [inline src/libs/SearchUtils.ts:474]: there is a tiny bit simpler way to create array from set
`const arr = [...(new Set())]` - I prefer this one because it uses destructuring and reads a bit better.

File: src/components/Search/types.ts [hunk 9]
```diff
@@ -49,7 +49,7 @@ type QueryFilter = {
     value: string | number;
 };
 
-type AdvancedFiltersKeys = ValueOf<typeof CONST.SEARCH.SYNTAX_FILTER_KEYS> | typeof CONST.SEARCH.SYNTAX_ROOT_KEYS.TYPE | typeof CONST.SEARCH.SYNTAX_ROOT_KEYS.STATUS;
+type AdvancedFiltersKeys = ValueOf<typeof CONST.SEARCH.SYNTAX_FILTER_KEYS>;
```
Comment (Kicu) [inline src/components/Search/types.ts:52]: great change 👏

**Review by Kicu — APPROVED** [review]
LGTM, left some more small comments

File: src/components/Search/SearchPageHeader.tsx [hunk 10]
```diff
@@ -316,7 +316,12 @@ function SearchPageHeader({queryJSON, hash, onSelectDeleteOption, setOfflineModa
             <Button
                 text={translate('search.filtersHeader')}
                 icon={Expensicons.Filters}
-                onPress={() => Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS)}
+                onPress={() => {
+                    const filters = SearchUtils.getFilters(queryJSON);
+                    const values = SearchUtils.getFiltersFormValues(filters, queryJSON);
+                    SearchActions.updateAdvancedFilters(values);
+                    Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS);
+                }}
```
Comment (luacmartins) [inline src/components/Search/SearchPageHeader.tsx:324]: NAB can we create a function for this logic

File: src/pages/Search/SearchTypeMenuNarrow.tsx [hunk 11]
```diff
@@ -96,7 +100,12 @@ function SearchTypeMenuNarrow({typeMenuItems, activeItemIndex, title}: SearchTyp
             </PressableWithFeedback>
             <Button
                 icon={Expensicons.Filters}
-                onPress={() => Navigation.navigate(ROUTES.SEARCH_ADVANCED_FILTERS)}
+                onPress={() => {
```
Comment (luacmartins) [inline src/pages/Search/SearchTypeMenuNarrow.tsx:103]: Same here

**Review by luacmartins — COMMENTED** [review]
Looks great. Left a small comment.

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/libs/SearchUtils.ts [hunk 12]
```diff
@@ -436,18 +436,20 @@ function getChatFiltersTranslationKey(has: ValueOf<typeof CONST.SEARCH.CHAT_TYPE
...
             if (keyInCorrectForm) {
                 return `${CONST.SEARCH.SYNTAX_FILTER_KEYS[keyInCorrectForm]}:${sanitizeString(filterValue as string)}`;
             }
         }
         if (filterKey === FILTER_KEYS.KEYWORD && filterValue) {
-            const keyInCorrectForm = (Object.keys(CONST.SEARCH.SYNTAX_FILTER_KEYS) as KeysOfFilterKeysObject[]).find((key) => CONST.SEARCH.SYNTAX_FILTER_KEYS[key] === filterKey);
-            if (keyInCorrectForm) {
-                return `${CONST.SEARCH.SYNTAX_FILTER_KEYS[keyInCorrectForm]}:${filterValue as string}`;
-            }
+            return `${filterValue as string}`;
```
[the remark this answers is not in the document]
Comment (289Adam289) [reply src/libs/SearchUtils.ts:445]: At first I din't think it was necessary but after some thought if we add operator symbols to `sanitizeString` function and sanitize every word of an input it should prevent some bugs.

[the remark this answers is not in the document]
Comment (289Adam289) [reply src/libs/SearchUtils.ts:445]: While working on keyword filter sanitization i came across another problem. Quotation marks are hard to parse. For example category with a single quote mark breaks the parser and even sanitizing it wont help. I suggest we discuss this problem in another issue cause it may needs some fundamental changes in grammar itself. cc @luacmartins @Kicu

**Comment by 289Adam289** [comment]
I don't think duplicate keyword filter values are a problem. If a user chooses to enter "test test" then I think we should leave it as it is.

File: src/libs/SearchUtils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (luacmartins) [reply src/libs/SearchUtils.ts]: I agree that it'd require changes to the grammar and we can handle that as a separate issue.

**Comment by luacmartins** [comment]
>I don't think duplicate keyword filter values are a problem. If a user chooses to enter "test test" then I think we should leave it as it is.

I agree this is not a problem. We're also updating the grammar [here](https://github.com/Expensify/App/pull/47789) that will combine them into a single node

**Review by luacmartins — DISMISSED** [review]
LGTM

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
@rayane-djouah all yours

**Review by rayane-d — APPROVED** [review]
LGTM and tests well

**Review by luacmartins — DISMISSED** [review]

**Comment by luacmartins** [comment]
@289Adam289 we got conflicts!

**Review by luacmartins — DISMISSED** [review]
LGTM

**Comment by luacmartins** [comment]
I'm gonna merge this PR as is and then work on a follow up to address [this comment](https://github.com/Expensify/App/pull/47690#issuecomment-2338627041)

**Comment by luacmartins** [comment]
@289Adam289 if you'd like you can remove the `Has` and `Is` filter from the advanced filters page, so we can move this PR forward. I'll work on adding them to the status bar, since that requires some backend changes.

**Comment by 289Adam289** [comment]
I can do that

**Comment by luacmartins** [comment]
It seems like we have a couple of minor bugs left. Let's squash them and get this one merged 😄

**Comment by Kicu** [comment]
To clarify about filter values in onyx.
Storing them in onyx was added only for the Advanced Filters Form to work.
In some of the filters we are using `FromProvider` which expects to get an onyx key with the form state - won't work otherwise. So it made the most sense to make the whole form use onyx as storage, both for single value filters and multi-picker filters.

However we store the values in onyx _only_ when the form (`/filters`) is open. When you're on other Search pages we store nothing in onyx - URL (or `params` in navigation state) are the single source of truth.
This is basically form-state

**Comment by 289Adam289** [comment]
Now Onyx is cleared when changing type. Reset button is only visible when filters are set and it clears only advanced filters.

**Review by rayane-d — APPROVED** [review]
LGTM and tests well

**Review by luacmartins — APPROVED** [review]
LGTM

File: src/libs/SearchUtils.ts [hunk 13]
```diff
@@ -612,6 +591,57 @@ function getFilters(queryJSON: SearchQueryJSON) {
...
+/**
+ * returns the values of the filters in a format that can be used in the SearchAdvancedFiltersForm as initial form values
+ */
+function getFiltersFormValues(queryJSON: SearchQueryJSON) {
+    const filters = getFilters(queryJSON);
+    const filterKeys = Object.keys(filters);
+    const filtersForm = {} as Partial<SearchAdvancedFiltersForm>;
+    for (const filterKey of filterKeys) {
+        if (filterKey === CONST.SEARCH.SYNTAX_FILTER_KEYS.REPORT_ID || filterKey === CONST.SEARCH.SYNTAX_FILTER_KEYS.MERCHANT || filterKey === CONST.SEARCH.SYNTAX_FILTER_KEYS.DESCRIPTION) {
+            filtersForm[filterKey] = filters[filterKey]?.[0]?.value.toString();
```
Comment (FitseTLT) [inline src/libs/SearchUtils.ts:603]: The filters array should be joined instead of using on the first value. Fixed [here](https://github.com/Expensify/App/pull/77414/changes)
